lp_solve_config/lp_solve/doc/en/lp_solve.1
.TH lp_solve 1 "version 5.5.0.10c"
.SH NAME
lp_solve \- command-line Mixed Integer Programming (MIP) solver
.SH SYNOPSIS
.B lp_solve
.RB [ options ]
.RB [ input_file ]
.SH DESCRIPTION
.B lp_solve
is a command-line interface to the lpsolve library, a Mixed Integer
Programming (MIP) linear-programming solver. The linear programming
models are read from a model file or can be read from standard
input. The format of the model should be in one of the formats MPS or
LP.

.P 
The program uses the simplex method for solving the system of linear
equations.
.SH OPTIONS

.TP
.TP
.B \-h 
Print a brief synopsis of 
.B lp_solve 
usage and options.

.TP
.B \-lp
read from an LP-format file. This is the default behavior.

.TP
.B \-mps
read from an MPS file in fixed format. 

.TP
.B \-fmps
read from an MPS file in free format. 

.TP
.BI "\-rxli xliname " filename
Read file with xli library.

.TP
.BI "\-rxlidata " datafilename
data file name for xli library.

.TP
.BI "\-rxliopt " options
options for xli library.

.TP
.BI "\-rbas " filename	
read basis from filename.

.TP
.BI "\-rpar " filename	
read parameters from filename.

.TP
.BI "\-rparopt " options
options for parameter file:
 -H headername: header name for parameters. By default 'Default'

.TP
.BI "\-wlp " filename	
write to LP file

.TP
.BI "\-wmps " filename	
write to an MPS file in fixed format

.TP
.BI "\-wfmps " filename	
write to an MPS file in free format

.TP
.BI "\-wxli xliname " filename
write file with xli library

.TP
.BI "\-wxliopt " options
options for xli library.

.TP
.BI "-wxlisol xliname " filename
write solution file with xli library

.TP
.BI "-wxlisolopt " options
options for xli library.

.TP
.BI "-wbas " filename	
write basis to filename.

.TP
.BI "-wpar " filename	
write parameters to filename.

.TP
.BI "-wparopt " options
options for parameter file:
 -H headername: header name for parameters. By default 'Default'

.TP
.B \-wafter
Write model after solve (useful if presolve used).

.TP
.B \-parse_only
parse input file but do not solve

.TP
.B \-nonames
Ignore variables and constraint names

.TP
.B \-norownames
Ignore constraint names

.TP
.B \-nocolnames
Ignore variable names

.TP
.B \-min
Minimize the LP problem. Using this option overrules any setting in
the input model file.

.TP
.B \-max
Maximize the LP problem. Using this option overrules any setting in
the input model file.

.TP
.BI "-r " value
Specify the maximum number of pivots between a re-inversion of the matrix.

.TP
.B "-piv" rule
Specify simplex pivot rule

.P
     -piv0: Select first
     -piv1: Select according to Dantzig
     -piv2: Select Devex pricing from Paula Harris (default)
     -piv3: Select steepest edge
.TP
.P
These pivot rules can be combined with any of the following:
.TP
     -pivf: In case of Steepest Edge, fall back to DEVEX in primal.
     -pivm: Multiple pricing.
     -piva: Temporarily use First Index if cycling is detected.
     -pivr: Adds a small randomization effect to the selected pricer.
     -pivll: Scan entering/leaving columns left rather than right.
     -pivla: Scan entering/leaving columns alternatingly left/right.
     -pivh: Use Harris' primal pivot logic rather than the default.
     -pivt: Use true norms for Devex and Steepest Edge initializations.

.TP
.B "-s"mode scaleloop
Use automatic problem scaling.
.P
     -s0: No scaling
     -s1: Geometric scaling (default)
     -s2: Curtis-reid scaling
     -s3: Scale to convergence using largest absolute value
     -s4: Numerical range-based scaling
     -s5: Scale to convergence using logarithmic mean of all values
     -s6: Scale based on the simple numerical range
     -s7: Scale quadratic
.TP
.P
These scaling rules can be combined with any of the following:
.TP
     -sp: also do power scaling.
     -si: also do Integer scaling (default).
     -se also do equilibration to scale to the -1..1 range (default).
.TP
.B \-presolve
presolve rows and columns of problem before start optimizing

.TP
.B \-presolverow
resolve rows of problem before start optimizing

.TP
.B \-presolvecol
resolve columns of problem before start optimizing

.TP
.B \-presolvel
Eliminate linearly dependent rows

.TP
.B \-presolves
Convert constraints to Special-ordered sets of Type 1 (SOS1)

.TP
.B \-presolver
Af the phase 1 solution process finds that a constraint is redundant,
then this constraint is deleted

.TP
.B \-presolvek
Simplification of knapsack-type constraints through addition of an
extra variable, which also helps bound the OF

.TP
.B \-presolveq
Direct substitution of one variable in 2-element equality
constraints; this requires changes to the constraint matrix
.TP
.B \-presolvem
Merge rows
.TP

.B \-presolvefd
COLFIXDUAL
.TP

.B \-presolvebnd
Presolve bounds

.TP
.B \-presolved
Presolve duals
.TP

.B \-presolvef
Identify implied free variables (releasing their explicit bounds)
.TP

.B \-presolveslk
IMPLIEDSLK

.TP
.B \-presolveg
Reduce (tighten) coefficients in integer models based on GCD argument

.TP
.B \-presolveb
Attempt to fix binary variables at one of their bounds

.TP
.B \-presolvec
Attempt to reduce coefficients in binary models

.TP
.B \-presolverowd
Idenfify and delete qualifying constraints that are dominated by
others, also fixes variables at a bound

.TP
.B \-presolvecold
Deletes variables (mainly binary), that are dominated by others (only
one can be non-zero)

TO BE CONTINUED....


lp_solve_config/CHANGES
10e to 10f
- Create a patches/dispute directory and move controversial or disputed patches
  there. Revert some of these disputed patches
- example -> test
- reinstate some of the examples in lp_example from old examples.

These are changes from the 5.5.0.10 release to 10e

- Various posted patches to the lpsolve mailing list.
- Add automake/autoconf (no libtool yet)
- Start a Unix manual page for lp_solve
- Add some regression tests

Patches from Andy Lutomirski in patches directory. All applied except
lpsolve_api_tweaks.patch.


lp_solve/5.5.2.0/.brew/lp_solve.rb
class NumpyHasHeaders < Requirement
  def numpy_include_dir
    "#{`python -c "import numpy.distutils.misc_util as u; print(u.get_numpy_include_dirs())[0]"`.strip}"
  end

  def satisfied?
    Dir["#{numpy_include_dir}/numpy/*.h"].any?
  end

  def fatal?
    true
  end

  def message; <<~EOS
    lp_solve requires NumPy headers not provided by Apple.
    Install numpy via the Homebrew/python formula:
      `brew tap homebrew/python && brew install numpy`
    EOS
  end
end

class LpSolve < Formula
  homepage "https://sourceforge.net/projects/lpsolve/"
  url "https://downloads.sourceforge.net/lpsolve/lp_solve_5.5.2.0_source.tar.gz"
  version "5.5.2.0" # automatic version parser spits out "solve" as version
  sha256 "5827a30b143105283f398a09419ea608719a2d7699ecea165a66d521803bcc9c"

  depends_on "python" => :optional
  depends_on NumpyHasHeaders.new if build.with? "python"

  resource "lp_solve_python" do
    # 'http://lpsolve.sourceforge.net/5.5/Python.htm'
    url "https://downloads.sourceforge.net/lpsolve/lp_solve_5.5.2.0_Python_source.tar.gz"
    sha256 "79683bc262e9da5b2fa7338ba190a9cd10559b9f2dbcb8f3bf07e2a81083ac51"
    version "5.5.2.0"
  end

  # Prefer OS X's fast BLAS implementation (patch stolen from fink-project)
  patch :DATA if OS.mac?

  def install
    # Thanks to superenv, we don't have to care if the ccc.osx build script
    # tells the compiler stupid things. And Xcode-only works like charm.

    cd "lpsolve55" do
      if OS.mac?
        system "sh", "ccc.osx", "#", "lpsolve55", "library"
        lib.install Dir["./bin/osx64/*.a"]
        lib.install Dir["./bin/osx64/*.dylib"]
      else
        system "sh", "ccc", "#", "lpsolve55", "library"
        lib.install Dir["./bin/ux64/*.a"]
        lib.install Dir["./bin/ux64/*.so"]
      end
    end

    cd "lp_solve" do
      if OS.mac?
        system "sh", "ccc.osx", "#", "lp_solve", "executable"
        bin.install "./bin/osx64/lp_solve"
      else
        system "sh", "ccc", "#", "lp_solve", "executable"
        bin.install "./bin/ux64/lp_solve"
      end
    end

    # Note, the demo does not build with lpsolve55. Upstream issue.

    include.install Dir["*.h"]
    include.install "shared/commonlib.h", "shared/myblas.h"
    include.install Dir["bfp/bfp_LUSOL/LUSOL/lusol*.h"]

    if build.with? "python"
      # In order to install into the Cellar, the dir must exist and be in the
      # PYTHONPATH.
      temp_site_packages = lib/which_python/"site-packages"
      mkdir_p temp_site_packages
      ENV["PYTHONPATH"] = temp_site_packages
      args = [
        "--force",
        "--verbose",
        "--install-scripts=#{share}/python",
        "--install-lib=#{temp_site_packages}",
        "--record=installed-files.txt",
      ]

      resource("lp_solve_python").stage do
        cd "extra/Python" do
          # On OS X malloc there is <sys/malloc.h> and <malloc/malloc.h>
          inreplace "hash.c", "#include <malloc.h>", "#include <sys/malloc.h>"
          # We know where the lpsolve55 lib is...
          inreplace "setup.py", "LPSOLVE55 = '../../lpsolve55/bin/ux32'", "LPSOLVE55 = '#{lib}'"
          # Correct path to lpsolve's include dir and go the official way to find numpy include_dirs
          inreplace "setup.py",
                    "include_dirs=['../..', NUMPYPATH],",
                    "include_dirs=['#{include}', '#{numpy_include_dir}'],"
          inreplace "setup.py", "(NUMPY, '1')", "('NUMPY', '1')"
          # Even their version number is broken ...
          inreplace "setup.py", 'version = "5.5.0.9",', "version = '#{version}',"

          system "python", "setup.py", "--no-user-cfg", "install", *args

          # Save the examples
          (share/"lp_solve").install Dir["ex*.py"], "lpdemo.py", "Python.htm"
        end
      end
    end
  end

  if build.with? "python"
    def numpy_include_dir
      "#{`python -c "import numpy.distutils.misc_util as u; print(u.get_numpy_include_dirs())[0]"`.strip}"
    end

    def which_python
      "python" + `python -c 'import sys;print(sys.version[:3])'`.strip
    end

    def caveats; <<~EOS
      For non-homebrew Python, you need to amend your PYTHONPATH like so:
        export PYTHONPATH=#{HOMEBREW_PREFIX}/lib/#{which_python}/site-packages:$PYTHONPATH

      Python examples and doc are installed to #{HOMEBREW_PREFIX}/share/lp_solve
      EOS
    end
  end

  test do
    input = <<~EOS
      max: 143 x + 60 y;

      120 x + 210 y <= 15000;
      110 x + 30 y <= 4000;
      x + y <= 75;
    EOS
    (testpath/"input.lp").write(input)
    output = `#{bin}/lp_solve -S3 input.lp`
    puts output
    match = output =~ /Value of objective function: 6315\.6250/
    raise if match.nil?
  end
end

__END__
diff --git a/lp_lib.h b/lp_lib.h
index 2ef654f..7b06ef8 100644
--- a/lp_lib.h
+++ b/lp_lib.h
@@ -104,7 +104,7 @@
 /* Specify use of the basic linear algebra subroutine library                */
 /* ------------------------------------------------------------------------- */
 #define libBLAS                  2        /* 0: No, 1: Internal, 2: External */
-#define libnameBLAS        "myBLAS"
+#define libnameBLAS        "/System/Library/Frameworks/Accelerate.framework/Frameworks/vecLib.framework/Versions/Current/libBLAS.dylib"
 
 
 /* Active inverse logic (default is optimized original etaPFI)               */
